key-rotator/manifest/manifest.go

```diff
@@ -103,9 +123,48 @@ func (m DataShareProcessorSpecificManifest) UpdateKeys(cfg UpdateKeysConfig) (Da
 	}
 	newM.PacketEncryptionKeyCSRs[kid] = newPEC
 
+	if err := validateUpdatedManifest(newM, m); err != nil {
+		return DataShareProcessorSpecificManifest{}, fmt.Errorf("manifest update validation error: %w", err)
+	}
 	return newM, nil
 }
 
+func validateUpdatedManifest(m, oldM DataShareProcessorSpecificManifest) error {
+	// Post-update, manifests must have at least one batch signing key version.
+	if len(m.BatchSigningPublicKeys) == 0 {
+		return errors.New("no batch signing public keys")
+	}
+
+	// Post-update, manifests must have exactly one packet encryption key version.
+	if len(m.PacketEncryptionKeyCSRs) != 1 {
+		return fmt.Errorf("expected exactly one packet encryption public key (had %d)", len(m.PacketEncryptionKeyCSRs))
+	}
+
+	// Post-update, manifests' non-key data must match pre-update manifest data exactly.
+	if !m.equalModuloKeys(oldM) {
+		return fmt.Errorf("non-key data modified")
+	}
+
+	// Post-update, manifests' key data for key versions that exist both pre- &
+	// post-update must match exactly.
+	for kid, key := range m.BatchSigningPublicKeys {
```

Comment:
Correct me if im wrong here, but I think there should be one more check of "these `kid`s should definitely exist on the new key version and error if it doesn't (?)

Reply:
Agreed: this was added as part of the follow-on PR https://github.com/abetterinternet/prio-server/pull/1136. (I didn't put them in this PR to keep PR sizes at least somewhat manageable, and because the validations in https://github.com/abetterinternet/prio-server/pull/1136 required refactoring a large number of tests in several packages.)